File: react-native/src/navigation.ts
/**
 * Minimal in-app screen router — no navigation library dependency needed for
 * this demo app. Push handlers (which run outside the React component tree,
 * including notifee background event handlers) call `navigate()`; `App.tsx`
 * subscribes to render the corresponding screen.
 */
export type ScreenName = 'home' | 'sale' | 'dashboard';

type Listener = (screen: ScreenName) => void;

let currentScreen: ScreenName = 'home';
const listeners = new Set<Listener>();

export function navigate(screen: ScreenName): void {
  currentScreen = screen;
  listeners.forEach((listener) => listener(screen));
}

export function getCurrentScreen(): ScreenName {
  return currentScreen;
}

/** Subscribes to screen changes; returns an unsubscribe function. */
export function subscribe(listener: Listener): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/**
 * Returns the in-app screen name for `deeplink` if it points at one of this
 * demo app's own screens (`zixflowdemo://sale`, `zixflowdemo://dashboard`),
 * or `null` if it should instead be treated as an external URL (opened via
 * `Linking.openURL`).
 */
export function resolveInAppRoute(deeplink?: string): ScreenName | null {
  if (!deeplink) return null;
  const match = /^zixflowdemo:\/\/(sale|dashboard)\b/i.exec(deeplink);
  if (!match) return null;
  return match[1].toLowerCase() as ScreenName;
}

File: flutter/lib/navigation.dart
import 'package:flutter/material.dart';

/// Global navigator key so push notification handlers (which run outside the
/// widget tree — including the background isolate used for FCM background
/// messages) can trigger in-app navigation when a Zixflow deeplink matches
/// one of this app's own screens.
final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

/// Route name for the demo "Flash Sale" screen, reachable via the
/// `zixflowdemo://sale` deeplink.
const String saleRoute = '/sale';

/// Route name for the demo "Dashboard" screen, reachable via the
/// `zixflowdemo://dashboard` deeplink.
const String dashboardRoute = '/dashboard';

/// Returns the named in-app route for [deeplink] if it points at one of this
/// demo app's own screens (`zixflowdemo://sale`, `zixflowdemo://dashboard`),
/// or `null` if it should instead be treated as an external URL (opened via
/// `url_launcher`).
String? resolveInAppRoute(String? deeplink) {
  if (deeplink == null || deeplink.isEmpty) return null;
  final uri = Uri.tryParse(deeplink);
  if (uri == null || uri.scheme != 'zixflowdemo') return null;

  switch (uri.host) {
    case 'sale':
      return saleRoute;
    case 'dashboard':
      return dashboardRoute;
    default:
      return null;
  }
}
